### PR TITLE
Improvements and fixes for file uploading. Prepares for handling of duplicate error.

### PR DESCRIPTION
**Pull request by bartkamphorst**
This PR makes way for implementing https://github.com/gollum/gollum/issues/1301.

### CODE REVIEW

**Comment by dometto** [comment]
Looks good to me!

1. Why is it only preparing for duplicate file handling? Looks pretty complete, but maybe I'm missing something.
2. some kind of feedback would be nice on success. An alert would be too intrusive, but perhaps some kind of visual response? @ProgramFan had some kind of visual response to the editor in his recent fix of the drag drop upload (before he implemented the spinner), perhaps we could do something like that for the dialog. But we can merge this without.
3. What about the possibility of uploading protected files? custom.css and such. Will not work when they're already present, obviously.

**Comment by bartkamphorst** [comment]
> Why is it only preparing for duplicate file handling? Looks pretty complete, but maybe I'm missing something.

It would be nice if we would return the user to a dialog to ask whether (s)he wants to overwrite the previous version of the file, and if so, resubmit the ajax request with an additional parameter or header that signals 'force overwrite'.

> some kind of feedback would be nice on success. An alert would be too intrusive, but perhaps some kind of visual response? @ProgramFan had some kind of visual response to the editor in his recent fix of the drag drop upload (before he implemented the spinner), perhaps we could do something like that for the dialog. But we can merge this without.

Agreed, but there wasn't any feedback before either.

> What about the possibility of uploading protected files? custom.css and such. Will not work when they're already present, obviously.

Same behavior as before. I'm treating that as a separate issue.

**Comment by dometto** [comment]
> Agreed, but there wasn't any feedback before either.

Fair enough, doesn't need to be part of this PR, but would be good to track the issue somewhere!

File: lib/gollum/templates/layout.mustache [hunk 1]
```diff
@@ -15,11 +15,15 @@
...
 
   <script>
-      var baseUrl = '{{base_url}}';
-      var uploadDest   = '{{upload_dest}}';
-      {{#page}}
-      var pageFullPath = '{{url_path}}';
-      {{/page}}
+	var baseUrl = '{{base_url}}';
+	var uploadDest = 'uploads';
+	var perPageUploads = '{{per_page_uploads}}';
```
Comment (dometto) [inline lib/gollum/templates/layout.mustache]: Wait, I'm not sure for the need for this now that the upload destination is being set using the referrer.

File: lib/gollum/views/layout.rb [hunk 2]
```diff
@@ -43,6 +43,10 @@ def css # custom css
       def js # custom js
         @js
       end
+      
+      def per_page_uploads
```
Comment (dometto) [inline lib/gollum/views/layout.rb]: If I'm right that the JS variable can go, this function should go too

File: lib/gollum/app.rb [hunk 3]
```diff
@@ -180,23 +180,33 @@ class App < Sinatra::Base
...
 
         if params[:file]
           fullname = params[:file][:filename]
           tempfile = params[:file][:tempfile]
         end
         halt 500 unless tempfile.is_a? Tempfile
 
-        # Remove page file dir prefix from upload path if necessary -- committer handles this itself
-        dir      = wiki.per_page_uploads ? params[:upload_dest] : ::File.join([wiki.page_file_dir, 'uploads'].compact)
+        if wiki.per_page_uploads
```
[the remark this answers is not in the document]
Comment (dometto) [reply lib/gollum/app.rb]: Just realized the setting of `@per_page_uploads` may be unnecessary, if the JS variable can indeed go and thus the views don't need to know about per page uploads. In that case, this line is fine as is, but the `@per_page_uploads` instance variable becomes superfluous.

**Comment by dometto** [comment]
I think this implements the functionality proposed in https://github.com/gollum/gollum/pull/1299, right? If so, that PR should be closed.

**Comment by bartkamphorst** [comment]
Correct!

File: lib/gollum/app.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (bartkamphorst) [reply lib/gollum/app.rb]: > Why not if @per_page_uploads, as you've just defined it above?

I like thinking of the instance variables as being set for use in the views, and having the code in the routes be "self-supporting" as it were. (I'll address your worry about the JS below).

File: lib/gollum/app.rb [hunk 4]
```diff
@@ -180,23 +180,33 @@ class App < Sinatra::Base
...
+          # remove base_url and gollum/* subpath if necessary
+          dir = request.referer.
+                  sub(request.base_url, '').
```
[the remark this answers is not in the document]
Comment (bartkamphorst) [reply lib/gollum/app.rb]: The referer includes protocol and base url (e.g., `http://localhost:4567/Home.md`) and subbing out the `request.base_url` (`e.g., http://localhost:4567`) gives me the URL path (i.e. `Home.md`). However, when using the drag/drop upload option, the path will often include part of the route, such as in `http://localhost:4567/gollum/edit/pathname/filename.md`. That's what the second substitution (`/.*gollum\/[-\w]+\//`) deals with.

File: lib/gollum/app.rb [hunk 5]
```diff
@@ -180,23 +180,33 @@ class App < Sinatra::Base
...
+                  sub(/.*gollum\/\w+\//, '')
+          # remove file extension 
+          dir = dir.sub(::File.extname(dir), '')
```
[the remark this answers is not in the document]
Comment (bartkamphorst) [reply lib/gollum/app.rb]: The former (referer is just the entire url). The edge case works correctly: uploads are stored in `uploads/Home.old/`.

File: lib/gollum/templates/layout.mustache (hunk 1 above)
Comment (bartkamphorst) [reply lib/gollum/templates/layout.mustache]: In the new setup the frontend and the backend determine the upload path separately. The frontend only uses this for showing/inserting the upload path, but it is no longer leading for determining the actual path (which happens in the backend). I like this approach better than having to set an instance variable for each route that possibly supports uploading (e.g., create, edit, show).

File: lib/gollum/views/layout.rb (hunk 2 above)
Comment (bartkamphorst) [reply lib/gollum/views/layout.rb]: This allows the JS to determine the correct upload path for use in the UI. See my previous comment for why I think this should stay.

File: lib/gollum/app.rb [hunk 6]
```diff
@@ -180,23 +180,33 @@ class App < Sinatra::Base
...
+          dir = ::File.join("uploads", dir)
+        else
+          # Remove page file dir prefix from upload path if necessary -- committer handles this itself
+          dir = ::File.join([wiki.page_file_dir, 'uploads'].compact)
```
[the remark this answers is not in the document]
Comment (bartkamphorst) [reply lib/gollum/app.rb]: No. If `wiki.page_file_dir == true` then `::File.join([wiki.page_file_dir, 'uploads'].compact)` will produce `[pagefiledir]/uploads`. When `wiki.page_file_dir != true` then `wiki.page_file_dir` actually is `nil`, and this is where the `compact` call does its magic by removing the `nil` entry from the array.

File: lib/gollum/app.rb [hunk 7]
```diff
@@ -180,23 +180,33 @@ class App < Sinatra::Base
...
+        end
+        halt 500 if dir.include?('..')
```
[the remark this answers is not in the document]
Comment (bartkamphorst) [reply lib/gollum/app.rb]: > if there is an equivalent to File::SEPARATOR we could use that.

I've been looking, but I haven't been able to find it. If you have more success finding the constant, I'll happily use it.

File: lib/gollum/templates/layout.mustache (hunk 1 above)
Comment (dometto) [reply lib/gollum/templates/layout.mustache]: > The frontend only uses this for showing/inserting the upload path

Ah, great! I didn't realize it was for that.

File: lib/gollum/views/layout.rb (hunk 2 above)
Comment (dometto) [reply lib/gollum/views/layout.rb]: 👍

**Pull request merged**